### PR TITLE
Support feel expressions for multi instance activities

### LINKED ISSUE #3804: Advanced expression support for multi-instance properties
**Is your feature request related to a problem? Please describe.**
For a multi-instance activity, I can define the `inputCollection` and the `outputElement` as FEEL expressions.

**Describe the solution you'd like**
Same as in #3792.

The `inputCollection` and the `outputElement` must be a FEEL expression and should not be static values (e.g. `foo`). But a FEEL expression that produces a static value is allowed (e.g. `=[1,2,3]`). 

**Additional context**
Currently, the `inputCollection` and the `outputElement` are expressions (i.e. variable reference or nested property of a variable `variable.property`). The issue is about changing the format of the expression to FEEL. 

The `inputElement` and `outputCollection` properties are no expressions. They just define the variable name where the value is stored.

### LINKED ISSUE #4101: Validate feel expressions for multi instance activities
**Is your feature request related to a problem? Please describe.**
We currently do not yet evaluate that a feel expression is used for the `inputCollection` and `outputElement` of multi instance activities.

### PR DESCRIPTION
**Pull request by korthout**
## Description

Specifically, this PR provides support for feel expressions for the
inputCollection and outputElement variables for multi instance
activities.

## Related issues

closes #3804 
closes #4101

### CODE REVIEW

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java [hunk 1]
```diff
@@ -44,12 +46,14 @@ protected void handleInnerActivity(final BpmnStepContext<ExecutableMultiInstance
 
     if (loopCharacteristics.isSequential()) {
 
-      final var array = readInputCollectionVariable(context).getSingleResult().getArray();
+      final var array =
+          readInputCollectionVariable(context).orElseThrow(IllegalStateException::new);
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: Throwing an exception here is dangerous :boom: It is very unlikely but the input collection variable might change. If it is not an array anymore then an exception is thrown and the workflow instance stops processing at all (i.e. added to the "blacklist").

It would be better to handle the case and create an incident instead. So, the workflow instance can be repaired and continue processing (which is not the case if it is added to the "blacklist").

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java [hunk 2]
```diff
@@ -69,35 +73,24 @@ protected boolean handleMultiInstanceBody(
     return true;
   }
 
-  private void updateOutputCollection(
+  protected void updateOutputCollection(
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: Why did you change the method visibility?

```suggestion
  private void updateOutputCollection(
```

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java [hunk 3]
```diff
@@ -69,35 +73,24 @@ protected boolean handleMultiInstanceBody(
...
-    final var elementVariable = readOutputElementVariable(context);
-    final var currentCollection = variablesState.getVariableLocal(bodyInstanceKey, variableName);
-
-    final var updatedCollection = insertAt(currentCollection, loopCounter, elementVariable);
-
-    variablesState.setVariableLocal(bodyInstanceKey, workflowKey, variableName, updatedCollection);
-  }
+    // the result needs to be cloned because readOutputElementVariable(Context) uses the same buffer
+    final var currentCollection =
+        cloneBuffer(variablesState.getVariableLocal(bodyInstanceKey, variableName));
```
Comment (saig0) [inline engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: We can be tricky and avoid cloning the buffer by moving `readOutputElementVariable(context);` on top :smiling_imp:  
The output element result is resolved by the expression language which uses its own buffers.

File: engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java [hunk 4]
```diff
@@ -589,13 +589,6 @@ public void shouldSetOutputElementVariable() {
     completeJobs(workflowInstanceKey, INPUT_COLLECTION.size());
 
     // then
-    assertThat(
-            RecordingExporter.variableRecords(VariableIntent.CREATED)
-                .withName(OUTPUT_ELEMENT_VARIABLE)
-                .limit(INPUT_COLLECTION.size()))
-        .extracting(r -> r.getValue().getValue())
-        .containsOnly("null");
-
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/multiinstance/MultiInstanceActivityTest.java]: I think we can keep this check now after dealing with the `outputElement` variable name.

File: expression-language/src/main/scala/io/zeebe/el/impl/feel/FeelEvaluationResult.scala [hunk 5]
```diff
@@ -7,12 +7,17 @@
...
+import java.{lang, util}
 
 import io.zeebe.el.{EvaluationResult, Expression, ResultType}
+import io.zeebe.util.buffer.BufferUtil
+import io.zeebe.util.buffer.BufferUtil.cloneBuffer
 import org.agrona.DirectBuffer
 import org.camunda.feel.syntaxtree.{Val, ValBoolean, ValContext, ValList, ValNull, ValNumber, ValString}
 
+import scala.collection.JavaConverters._
+import scala.collection.mutable.ListBuffer
```
Comment (saig0) [inline expression-language/src/main/scala/io/zeebe/el/impl/feel/FeelEvaluationResult.scala]: Optimized imports by Idea:

```suggestion
import java.{lang, util}

import io.zeebe.el.{EvaluationResult, Expression, ResultType}
import io.zeebe.util.buffer.BufferUtil.cloneBuffer
import org.agrona.DirectBuffer
import org.camunda.feel.syntaxtree.{Val, ValBoolean, ValContext, ValList, ValNull, ValNumber, ValString}

import scala.jdk.CollectionConverters._
```

**Review by saig0 — CHANGES_REQUESTED** [review]
Looks good so far :+1: 

What is currently missing:
* a validation that the expressions are valid FEEL expressions (see `ZeebeRuntimeValidators`) 
  * it still checks that they are "JSON path" expressions
  * this avoids that I can use FEEL expressions
* a test to verify that an incident is raised and can be resolved if the `outputElement` expression fails (see `MultiInstanceIncidentTest`)
* a test to verify that the `EvaluationResult` can return a list (see `io.zeebe.el.EvaluationResultTest#listExpression()`)

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java (hunk 1 above)
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: @saig0 Ah okay, that's a good reason to avoid throwing an exception here. But how should we deal with the situation then? In case the variable is no longer an array, an incident is already raised inside the `readInputCollectionVariable` method. So we don't have to do that here. However, should we immediately return from this `handleInnerActivity` when we see the optional to be empty or should the rest of the method still be executed? For example, do we still need to call `super.handleInnerActivity`?

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java (hunk 2 above)
Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: oops 😄

File: engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java (hunk 1 above)
Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: :+1: We should return `false` immediately. Otherwise, we may continue the workflow instance. 

I have more and more the feeling that this API needs an `Either` return type...

Comment (korthout) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: @saig0 I'm all for it. Here are some implementations for Java: https://stackoverflow.com/questions/26162407/is-there-an-equivalent-of-scalas-either-in-java-8. The only issue I see with introducing one to our code is consistency. I think the ExpressionProcessor API should be migrated to using it in one PR, to make sure how we use it in the same way everywhere. That might take some effort.

Comment (saig0) [reply engine/src/main/java/io/zeebe/engine/processor/workflow/handlers/multiinstance/MultiInstanceBodyCompletedHandler.java]: :+1:

**Comment by korthout** [comment]
@saig0 I've made some additional changes.

> Looks good so far 👍
> 
> What is currently missing:
> 
> * ~~a validation that the expressions are valid FEEL expressions (see `ZeebeRuntimeValidators`)~~
>   
>   * ~~it still checks that they are "JSON path" expressions~~
>   * ~~this avoids that I can use FEEL expressions~~
> * ~~a test to verify that an incident is raised and can be resolved if the `outputElement` expression fails (see `MultiInstanceIncidentTest`)~~
> * ~~a test to verify that the `EvaluationResult` can return a list (see `io.zeebe.el.EvaluationResultTest#listExpression()`)~~

Please have a look at the test in `MultiInstanceIncidentTest`. I'd like to improve its performance, but I don't yet see how. Currently its about 10s to run which is way too long for such a small test.

File: engine/src/test/java/io/zeebe/engine/processor/workflow/incident/MultiInstanceIncidentTest.java [hunk 6]
```diff
@@ -115,9 +128,61 @@ public void shouldCreateIncidentIfInputVariableIsNotAnArray() {
...
+
+    for (final long key : jobs.getValue().getJobKeys()) {
+      ENGINE.job().withKey(key).complete();
+    }
+    final List<Record<JobRecordValue>> completed =
+        RecordingExporter.jobRecords(JobIntent.COMPLETED)
+            .withType(jobType)
+            .withWorkflowInstanceKey(workflowInstanceKey)
+            .asList();
+    assertThat(completed).hasSize(3);
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/incident/MultiInstanceIncidentTest.java]: In this test, we need to complete only one job. This can be done with the shortcut:

```suggestion
    ENGINE.job().withType(jobType).ofInstance(workflowInstanceKey).complete();
```

File: engine/src/test/java/io/zeebe/engine/processor/workflow/incident/MultiInstanceIncidentTest.java [hunk 7]
```diff
@@ -115,9 +128,61 @@ public void shouldCreateIncidentIfInputVariableIsNotAnArray() {
...
+            .withElementId(ELEMENT_ID)
+            .getFirst();
+
+    Assertions.assertThat(incident.getValue())
+        .hasElementInstanceKey(elementInstance.getKey())
+        .hasElementId(elementInstance.getValue().getElementId())
+        .hasErrorType(ErrorType.EXTRACT_VALUE_ERROR)
+        .hasErrorMessage(
+            "failed to parse expression 'some_result': [1.12] error: "
+                + "'in' expected but end of source found\n\nsome_result\n           ^");
```
Comment (saig0) [inline engine/src/test/java/io/zeebe/engine/processor/workflow/incident/MultiInstanceIncidentTest.java]: This is not the failure you're looking for :robot: 

It seems that the FEEL engine doesn't like variable names starting with `some` (which is a reserved word in FEEL). 

Since the output element variable would be created when the multi-instance element is activated, we need to use some kind of expression (e.g. `sum(r)`).

**Review by saig0 — CHANGES_REQUESTED** [review]

**Review by saig0 — APPROVED** [review]
:+1:
